File: src/main/java/Back/bailAutoComplet/BailAutoComplet/Service/IrlService.java
package Back.bailAutoComplet.BailAutoComplet.Service;

import Back.bailAutoComplet.BailAutoComplet.Dto.IrlDto;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.NodeList;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import java.io.ByteArrayInputStream;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.time.Instant;
import java.util.Optional;

/**
 * Récupère la dernière valeur de l'IRL (Indice de Référence des Loyers) publiée
 * par l'INSEE via le service web SDMX public de la BDM (Banque de Données
 * Macroéconomiques). Aucune clé API n'est nécessaire pour cet endpoint.
 *
 * La valeur est mise en cache en mémoire : l'IRL n'étant publié qu'une fois par
 * trimestre, on évite d'interroger l'INSEE à chaque appel.
 */
@Service
public class IrlService {

    private static final Logger log = LoggerFactory.getLogger(IrlService.class);

    @Value("${insee.bdm.base-url:https://www.bdm.insee.fr/series/sdmx/data/SERIES_BDM}")
    private String baseUrl;

    @Value("${insee.irl.idbank:001515333}")
    private String idbank;

    @Value("${insee.irl.cache-ttl-hours:6}")
    private long cacheTtlHours;

    private final HttpClient httpClient = HttpClient.newBuilder()
            .connectTimeout(Duration.ofSeconds(10))
            .build();

    private volatile IrlDto cachedIrl;
    private volatile Instant cachedAt;

    /**
     * @return la dernière valeur IRL disponible, ou {@link Optional#empty()} si
     * l'INSEE est injoignable ou la réponse illisible (l'appelant conserve
     * alors la valeur déjà stockée).
     */
    public Optional<IrlDto> getLatestIrl() {
        IrlDto cached = cachedIrl;
        if (cached != null && cachedAt != null
                && cachedAt.isAfter(Instant.now().minus(Duration.ofHours(cacheTtlHours)))) {
            return Optional.of(cached);
        }

        try {
            Optional<IrlDto> fresh = fetchLatestIrl();
            fresh.ifPresent(irl -> {
                cachedIrl = irl;
                cachedAt = Instant.now();
            });
            // En cas d'échec ponctuel, on renvoie l'éventuelle valeur en cache (même expirée).
            return fresh.isPresent() ? fresh : Optional.ofNullable(cached);
        } catch (Exception e) {
            log.warn("Impossible de récupérer l'IRL depuis l'INSEE : {}", e.getMessage());
            return Optional.ofNullable(cached);
        }
    }

    private Optional<IrlDto> fetchLatestIrl() throws Exception {
        // On limite la fenêtre aux 3 dernières années pour une réponse légère.
        String url = baseUrl + "/" + idbank + "?startPeriod=" + (Instant.now().atZone(java.time.ZoneOffset.UTC).getYear() - 2);

        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .timeout(Duration.ofSeconds(15))
                .header("Accept", "application/xml")
                .GET()
                .build();

        HttpResponse<byte[]> response = httpClient.send(request, HttpResponse.BodyHandlers.ofByteArray());
        if (response.statusCode() != 200) {
            log.warn("L'INSEE a répondu avec le code HTTP {}", response.statusCode());
            return Optional.empty();
        }

        return parseLatestObservation(response.body());
    }

    /**
     * Parse la réponse SDMX (StructureSpecificData) et retourne l'observation la
     * plus récente. Les {@code <Obs>} portent les attributs {@code TIME_PERIOD}
     * (ex. "2026-Q1") et {@code OBS_VALUE} (ex. "146.6").
     */
    private Optional<IrlDto> parseLatestObservation(byte[] body) throws Exception {
        DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
        // Durcissement anti-XXE : on ne lit que des données, pas d'entités externes.
        factory.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true);
        factory.setNamespaceAware(true);
        DocumentBuilder builder = factory.newDocumentBuilder();
        Document doc = builder.parse(new ByteArrayInputStream(body));

        NodeList obsNodes = doc.getElementsByTagNameNS("*", "Obs");
        String bestPeriod = null;
        String bestValue = null;
        for (int i = 0; i < obsNodes.getLength(); i++) {
            Element obs = (Element) obsNodes.item(i);
            String period = obs.getAttribute("TIME_PERIOD");
            String value = obs.getAttribute("OBS_VALUE");
            if (period.isEmpty() || value.isEmpty()) {
                continue;
            }
            // Le format "YYYY-Qn" se trie correctement en ordre lexicographique.
            if (bestPeriod == null || period.compareTo(bestPeriod) > 0) {
                bestPeriod = period;
                bestValue = value;
            }
        }

        if (bestPeriod == null) {
            log.warn("Aucune observation IRL trouvée dans la réponse de l'INSEE");
            return Optional.empty();
        }

        return Optional.of(new IrlDto(bestValue, formatTrimestre(bestPeriod)));
    }

    /**
     * Transforme une période SDMX "2026-Q1" en libellé de trimestre "T1 2026".
     */
    private String formatTrimestre(String timePeriod) {
        String[] parts = timePeriod.split("-Q");
        if (parts.length == 2) {
            return "T" + parts[1] + " " + parts[0];
        }
        return timePeriod;
    }
}

File: src/irl/irl.service.ts
import { Injectable, Logger } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { XMLParser } from 'fast-xml-parser';
import { IrlDto } from './dto/irl.dto';

/**
 * Récupère la dernière valeur de l'IRL (Indice de Référence des Loyers) publiée
 * par l'INSEE via le service web SDMX public de la BDM (Banque de Données
 * Macroéconomiques). Aucune clé API n'est nécessaire pour cet endpoint.
 *
 * La valeur est mise en cache en mémoire : l'IRL n'étant publié qu'une fois par
 * trimestre, on évite d'interroger l'INSEE à chaque appel.
 */
@Injectable()
export class IrlService {
  private readonly logger = new Logger(IrlService.name);

  private readonly baseUrl: string;
  private readonly idbank: string;
  private readonly cacheTtlHours: number;

  private cachedIrl: IrlDto | null = null;
  private cachedAt: number | null = null;

  constructor(configService: ConfigService) {
    this.baseUrl = configService.get<string>(
      'INSEE_BDM_BASE_URL',
      'https://www.bdm.insee.fr/series/sdmx/data/SERIES_BDM',
    );
    this.idbank = configService.get<string>('INSEE_IRL_IDBANK', '001515333');
    this.cacheTtlHours = Number(
      configService.get<string>('INSEE_IRL_CACHE_TTL_HOURS', '6'),
    );
  }

  /**
   * @returns la dernière valeur IRL disponible, ou `null` si l'INSEE est
   * injoignable ou la réponse illisible (l'appelant conserve alors la valeur
   * déjà stockée en base).
   */
  async getLatestIrl(): Promise<IrlDto | null> {
    const cached = this.cachedIrl;
    const ttlMs = this.cacheTtlHours * 60 * 60 * 1000;

    if (
      cached &&
      this.cachedAt !== null &&
      Date.now() - this.cachedAt < ttlMs
    ) {
      return cached;
    }

    try {
      const fresh = await this.fetchLatestIrl();

      if (fresh) {
        this.cachedIrl = fresh;
        this.cachedAt = Date.now();
        return fresh;
      }

      // En cas d'échec ponctuel, on renvoie l'éventuelle valeur en cache
      // (même expirée).
      return cached;
    } catch (error) {
      this.logger.warn(
        `Impossible de récupérer l'IRL depuis l'INSEE : ${
          error instanceof Error ? error.message : String(error)
        }`,
      );
      return cached;
    }
  }

  private async fetchLatestIrl(): Promise<IrlDto | null> {
    // On limite la fenêtre aux 3 dernières années pour une réponse légère.
    const startPeriod = new Date().getUTCFullYear() - 2;
    const url = `${this.baseUrl}/${this.idbank}?startPeriod=${startPeriod}`;

    const response = await fetch(url, {
      headers: { Accept: 'application/xml' },
      signal: AbortSignal.timeout(15_000),
    });

    if (!response.ok) {
      this.logger.warn(
        `L'INSEE a répondu avec le code HTTP ${response.status}`,
      );
      return null;
    }

    return this.parseLatestObservation(await response.text());
  }

  /**
   * Parse la réponse SDMX (StructureSpecificData) et retourne l'observation la
   * plus récente. Les `<Obs>` portent les attributs `TIME_PERIOD`
   * (ex. « 2026-Q1 ») et `OBS_VALUE` (ex. « 146.6 »).
   */
  private parseLatestObservation(body: string): IrlDto | null {
    // Durcissement anti-XXE : on ne lit que des données, pas de DTD.
    if (/<!DOCTYPE/i.test(body)) {
      this.logger.warn(
        'Réponse INSEE rejetée : elle contient une déclaration DOCTYPE',
      );
      return null;
    }

    const parser = new XMLParser({
      ignoreAttributes: false,
      attributeNamePrefix: '',
      removeNSPrefix: true,
    });

    const observations = this.collectObservations(parser.parse(body));

    let bestPeriod: string | null = null;
    let bestValue: string | null = null;

    for (const observation of observations) {
      const period = observation.TIME_PERIOD;
      const value = observation.OBS_VALUE;

      if (!period || !value) {
        continue;
      }

      // Le format « YYYY-Qn » se trie correctement en ordre lexicographique.
      if (bestPeriod === null || period > bestPeriod) {
        bestPeriod = period;
        bestValue = value;
      }
    }

    if (bestPeriod === null || bestValue === null) {
      this.logger.warn(
        "Aucune observation IRL trouvée dans la réponse de l'INSEE",
      );
      return null;
    }

    return { valIrl: bestValue, tIrl: this.formatTrimestre(bestPeriod) };
  }

  /**
   * Les `<Obs>` sont imbriquées à une profondeur variable selon le format SDMX
   * renvoyé : on parcourt donc tout l'arbre plutôt que de figer un chemin.
   */
  private collectObservations(
    node: unknown,
    accumulator: Array<Record<string, string>> = [],
  ): Array<Record<string, string>> {
    if (Array.isArray(node)) {
      for (const item of node) {
        this.collectObservations(item, accumulator);
      }
      return accumulator;
    }

    if (node === null || typeof node !== 'object') {
      return accumulator;
    }

    for (const [key, value] of Object.entries(node)) {
      if (key === 'Obs') {
        const observations = Array.isArray(value) ? value : [value];
        for (const observation of observations) {
          if (observation && typeof observation === 'object') {
            accumulator.push(observation as Record<string, string>);
          }
        }
        continue;
      }

      this.collectObservations(value, accumulator);
    }

    return accumulator;
  }

  /** Transforme une période SDMX « 2026-Q1 » en libellé de trimestre « T1 2026 ». */
  private formatTrimestre(timePeriod: string): string {
    const parts = timePeriod.split('-Q');
    return parts.length === 2 ? `T${parts[1]} ${parts[0]}` : timePeriod;
  }
}
